### PR TITLE
feat: Add multiprocessing

### PR DESCRIPTION
**Pull request by Pringled**
Some simple code for benchmarking:

```python
from time import perf_counter

from model2vec import StaticModel
from datasets import load_dataset

def main(use_multiprocessing: bool):
    ds = load_dataset(
        "wikimedia/wikipedia",  
        data_files="20231101.en/train-00000-of-00041.parquet"
    )["train"]
    texts = ds["text"]

    model = StaticModel.from_pretrained("minishlab/potion-base-8M")

    start = perf_counter()
    output = model.encode(
        sentences=texts, 
        show_progress_bar=True, 
        use_multiprocessing=use_multiprocessing,
    )
    total_time = perf_counter() - start

    docs_per_second = len(texts) / total_time
    print(f"Processed {len(texts)} docs in {total_time:.3f}s.")
    print(f"Docs per second: {docs_per_second:.2f}")
    print("Output shape:", output.shape)


if __name__ == "__main__":
    print("Multiprocessing=False")
    main(use_multiprocessing=False)

    print("\nMultiprocessing=True")
    main(use_multiprocessing=True)
```

On my machine, this gives:

```
Multiprocessing=False
Processed 156289 docs in 34.067s.
Docs per second: 4587.69
Output shape: (156289, 256)
```

```
Multiprocessing=True
Processed 156289 docs in 10.407s.
Docs per second: 15017.26
Output shape: (156289, 256)
```

So roughly 3x faster.

### CODE REVIEW

**Comment by Pringled** [comment]
Threshold of number of samples after which to apply MP is currently set to 10k based on the following plot (I think we can keep a safe margin since the differences are relatively small around the inflection point of ~6k samples):

Screenshot 2024-12-27 at 17 06 41

File: model2vec/model.py [hunk 1]
```diff
@@ -1,22 +1,25 @@
...
 
-from model2vec.utils import load_local_model
+from model2vec.utils import ProgressParallel, load_local_model
 
 PathLike = Union[Path, str]
 
-
 logger = getLogger(__name__)
 
+MULTIPROCESSING_THRESHOLD = 10_000  # Minimum number of sentences to use multiprocessing
```
Comment (stephantul) [inline model2vec/model.py:21]: maybe make this private

File: model2vec/model.py [hunk 2]
```diff
@@ -186,24 +190,40 @@ def encode_as_sequence(
             If this is None, no truncation is done.
         :param batch_size: The batch size to use.
         :param show_progress_bar: Whether to show the progress bar.
+        :param use_multiprocessing: Whether to use multiprocessing.
```
Comment (stephantul) [inline model2vec/model.py:193]: if you want you can add additional information about constraints here.

File: model2vec/model.py [hunk 3]
```diff
@@ -237,6 +258,7 @@ def encode(
         :param max_length: The maximum length of the sentences. Any tokens beyond this length will be truncated.
             If this is None, no truncation is done.
         :param batch_size: The batch size to use.
+        :param use_multiprocessing: Whether to use multiprocessing.
```
Comment (stephantul) [inline model2vec/model.py:261]: idem on comment above

**Review by stephantul — APPROVED** [review]
Veryyyy nice.

File: model2vec/model.py (hunk 1 above)
Comment (Pringled) [reply model2vec/model.py:21]: done!

File: model2vec/model.py (hunk 2 above)
Comment (Pringled) [reply model2vec/model.py:193]: Good point, will update

**Pull request merged**